File: scala/src/test/scala/mtl/mtl.scala
package mtl

import cats.data.{ReaderT, Writer}
import Example.*
import org.scalatest.wordspec.AnyWordSpec
import org.scalatest.matchers.should.Matchers

// We can define a custom F to test the core logic of our function. Here we can use
// a specific monad stack: a Reader will allow us to mock the coin flip result
// a Writer will allow us to log the console output
type TestApp[A] = ReaderT[Writer[String, _], Boolean, A]

extension [A](a: TestApp[A])
  def runWithRiggedCoin(coinResult: Boolean): (String, A) =
    a.run(coinResult).run

given CoinFlip[TestApp] with
  def flipCoin: TestApp[Boolean] = ReaderT.ask

given Console[TestApp] with
  def printLine(s: String): TestApp[Unit] = ReaderT(_ => Writer.tell(s))

// Now testing the core logic is a breeze!
class Test extends AnyWordSpec, Matchers:
  val program = maybeDouble[TestApp](10)
  "maybeDouble" when {
    "executed" should {
      "always print a message regardless of the coin flip result" in {
        program.runWithRiggedCoin(true)._1 shouldBe "Flipping a coin!"
        program.runWithRiggedCoin(false)._1 shouldBe "Flipping a coin!"
      }
    }
    "the coin flip result is heads" should {
      "double the number" in {
        program.runWithRiggedCoin(true)._2 shouldBe 20
      }
    }
    "the coin flip result is tails" should {
      "not double the number" in {
        program.runWithRiggedCoin(false)._2 shouldBe 10
      }
    }
  }

File: haskell/tests/Test/MTL.hs
module Test.MTL (test) where

import Control.Monad.Trans.Class (lift)
import Control.Monad.Trans.Reader (ReaderT (runReaderT), ask)
import Control.Monad.Trans.Writer (Writer, runWriter, tell)
import MTL (CoinFlip (..), Console (..), maybeDouble)
import Test.Hspec (SpecWith, context, describe, it, shouldBe)

newtype TestApp a = TestApp (ReaderT Bool (Writer String) a) deriving (Functor, Applicative, Monad)

runWithRiggedCoin :: TestApp a -> Bool -> (a, String)
runWithRiggedCoin (TestApp app) coinResult = runWriter $ runReaderT app coinResult

instance CoinFlip TestApp where
  flipCoin :: TestApp Bool
  flipCoin = TestApp ask

instance Console TestApp where
  printLine :: String -> TestApp ()
  printLine = TestApp . lift . tell

test :: SpecWith ()
test = do
  let program = maybeDouble 10
  describe "maybeDouble" $ do
    context "when executed" $ do
      it "should output a message" $ do
        let (_, msg1) = program `runWithRiggedCoin` True
        let (_, msg2) = program `runWithRiggedCoin` False
        msg1 `shouldBe` "Flipping a coin!"
        msg2 `shouldBe` "Flipping a coin!"

    context "when the coin flip result is heads" $ do
      it "should double the number" $ do
        let (res, _) = program `runWithRiggedCoin` True
        res `shouldBe` 20

    context "when the coin flip result is tails" $ do
      it "should not double the number" $ do
        let (res, _) = program `runWithRiggedCoin` False
        res `shouldBe` 10
